### DISCUSSION TITLE
Using MikroORM with TypeDI

### DISCUSSION THREAD

**Discussion by gradinarot** [mikro-orm/mikro-orm#1828] [Q&A]
How can I verify that EntityManager was properly forked since I'm using it with TypeDI?

**Comment by B4nan** [comment]
As long as you are using `RequestContext` helper, you should be fine. The DI (any library, not just TypeDI) will maintain the global instances only. You are the one, who create forks (inside some middleware, where you call `RequestContext.create()`). The appropriate fork is then used automatically even if you work with the global instance. 

You can verify the contextual EM ID via `em.getContext().id` - if this ID is different for each request, you got it right.

**Reply by gradinarot** [reply]
This is the way it works for me, you think this is ok. I'm useing TypeGraphql together with Mikroorm. The buildSchema func is from type-graphql
![image](https://user-images.githubusercontent.com/17768157/118840821-0ada9800-b8c8-11eb-8ae3-3db1dee45cd8.png)

![image](https://user-images.githubusercontent.com/17768157/118840999-2e054780-b8c8-11eb-9839-90591bdebd59.png)

And in the end I'm destroying that container 
![image](https://user-images.githubusercontent.com/17768157/118841286-791f5a80-b8c8-11eb-8db8-0f26ffe3849e.png)

**Reply by B4nan** [reply]
Never used TypeDI, but it looks fine to me. As mentioned, you could use the `RequestContext` helper that will handle this for you behind the scenes, automatically, so you can use the global instances in DI. Your code looks like you are creating the forks explicitly, which is totally fine, just a bit more verbose (and you will end up with more class instances in memory). 

`RequestContext` works via domain API (or `AsyncLocalStorage` in v5) and provides a scope for the whole request, where we can take the right instances from any place (via static method, without changing any code). I know there were some issues with apollo as it registers body parser too early, so if you try it, be sure to regiser body parser yourself as described here https://mikro-orm.io/docs/next/usage-with-nestjs/#request-scoping-when-using-graphql

**Reply by firmanjml** [reply]
> This is the way it works for me, you think this is ok. I'm useing TypeGraphql together with Mikroorm. The buildSchema func is from type-graphql
> ![image](https://user-images.githubusercontent.com/17768157/118840821-0ada9800-b8c8-11eb-8ae3-3db1dee45cd8.png)
> 
> ![image](https://user-images.githubusercontent.com/17768157/118840999-2e054780-b8c8-11eb-9839-90591bdebd59.png)
> 
> And in the end I'm destroying that container
> ![image](https://user-images.githubusercontent.com/17768157/118841286-791f5a80-b8c8-11eb-8db8-0f26ffe3849e.png)

Hi @gradinarot!
I was wondering, is it safe to loop through the Metadata from the EM to get the Entities and set the entity to the container rather than manually hardcoded it?

![image](https://user-images.githubusercontent.com/37127022/122646056-f18b6c80-d14f-11eb-9c15-1f64cdfafbec.png)

**Comment by gradinarot** [comment]
Tried both but it didn't work for me. For some reason storage was returning undefined id
